controlplane/kubeadm/controllers/controller.go

```diff
@@ -284,14 +284,14 @@ func (r *KubeadmControlPlaneReconciler) reconcile(ctx context.Context, cluster *
 	conditions.SetAggregate(controlPlane.KCP, controlplanev1.MachinesReadyCondition, ownedMachines.ConditionGetters(), conditions.AddSourceRef())
 
 	// Control plane machines rollout due to configuration changes (e.g. upgrades) takes precedence over other operations.
-	needRollout := controlPlane.MachinesNeedingRollout()
+	needRollout := r.MachinesNeedingRollout(ctx, controlPlane)
```

Comment:
I think that this change will benefit from continuing to use `controlplane` as a source for `MachinesNeedingRollout` (e.g. this will avoid changes in methods signatures for `upgradeControlPlane`,  `scaleDownControlPlane`) 
Have you considered passing `r.Client` and `ctx` as input to `NewControlPlane`?

Reply:
I think we should call MachinesNeedingRollout() only in one place.
Calling it multiple places may result in inconsistent set of machines to be upgraded. So, to avoid calling this method in selectMachineForScaleDown, we need to change the signature. Wdyt? 

cc @vincepri

Reply:
In the future we might have to rework the codebase to be consistent around the usage of different structs, is it ok to do this later @fabriziopandini ?

Reply:
It is ok if there is a plan to rework. 
(I just added https://github.com/kubernetes-sigs/cluster-api/issues/3248#issuecomment-650344463 on the same line)